axelor-account/src/main/java/com/axelor/apps/account/service/payment/invoice/payment/InvoicePaymentCreateServiceImpl.java

```diff
@@ -201,49 +202,90 @@ protected void computeAdvancePaymentImputation(InvoicePayment invoicePayment, Mo
    * @return the found advance invoice if the move is from a payment that comes from this invoice.
    *     null in other cases
    */
-  protected Invoice determineIfReconcileFromInvoice(Move move) {
+  protected Invoice determineIfReconcileFromInvoice(Move move, int operationTypeSelect) {
     List<MoveLine> moveLineList = move.getMoveLineList();
     if (moveLineList == null || moveLineList.size() != 2) {
       return null;
     }
     InvoicePaymentRepository invoicePaymentRepo = Beans.get(InvoicePaymentRepository.class);
     for (MoveLine moveLine : moveLineList) {
       // search for the reconcile between the debit line
-      if (moveLine.getDebit().compareTo(BigDecimal.ZERO) > 0) {
-        Reconcile reconcile =
-            Beans.get(ReconcileRepository.class)
-                .all()
-                .filter("self.debitMoveLine = ?", moveLine)
-                .fetchOne();
-        if (reconcile == null) {
-          return null;
+      if (operationTypeSelect == InvoiceRepository.OPERATION_TYPE_SUPPLIER_PURCHASE) {
+        if (moveLine.getCredit().compareTo(BigDecimal.ZERO) > 0) {
+          Reconcile reconcile = findReconcileFromMoveLine(moveLine, false);
+          if (reconcile == null) {
+            return null;
+          }
+          // in the reconcile, search for the credit line to get the
+          // associated payment
+          MoveLine canditateMoveLine = reconcile.getDebitMoveLine();
+          if (canditateMoveLine == null || canditateMoveLine.getMove() == null) {
+            continue;
+          }
+          Move candidatePaymentMove = canditateMoveLine.getMove();
+          InvoicePayment invoicePayment =
+              invoicePaymentRepo
+                  .all()
+                  .filter("self.move = :_move")
+                  .bind("_move", candidatePaymentMove)
+                  .fetchOne();
+          // if the invoice linked to the payment is an advance
+          // payment, then return true.
+          if (invoicePayment != null
+              && invoicePayment.getInvoice() != null
+              && invoicePayment.getInvoice().getOperationSubTypeSelect()
+                  == InvoiceRepository.OPERATION_SUB_TYPE_ADVANCE) {
+            return invoicePayment.getInvoice();
```

Comment:
to refactor into findAdvancePaymentInvoiceFromPaymentMoveLine(MoveLine moveLine, int operationTypeSelect)

Reply:
done